### PR TITLE
Always make everything black if camera is within node

### LINKED ISSUE #12429: Seeing through blocks glitch
##### Minetest version
```
5.5.1
```

##### OS / Hardware

Operating system: Windows
CPU: Intel Core i5

GPU model: Intel HD Graphics Family
OpenGL version: Not sure

##### Summary
Players can x-ray by standing inside a block at the edge

##### Steps to reproduce
1. Get at least two blocks
2. Get near a wall
3. Place the blocks inside you
4. Move to the edge of the blocks slowly, but don't go outside of the blocks
5. At one point, you will be able to x-ray

### PR DESCRIPTION
**Pull request by appgurueu**
WIP. Closes https://github.com/minetest/minetest/issues/12429.

Simply checks all nodes within a +/- near clipping plane cuboid for solidness. Technically a sphere should be used, but the cuboid is way easier to implement and is only off by less than the very slim near plane ($\sqrt{3} - 1$ to be precise) at the *literal corner cases* which can safely be ignored.

Furthermore this makes no distinction for 3rd person view, as that would make it trivial to circumvent this fix (in fact I only knew the reported "glitch" from 3rd person, where no black color is applied for solid nodes).

# How to test

Go inside nodes as described in the issue. Now move your camera *just outside* the nodes such that the node face is clipped, allowing you to see through the node ("X-ray"):

screenshot_20220622_233202

Now apply the patch from this PR however you please and recompile. When you now rejoin, your screen should be all black:

screenshot_20220622_233443

Switch to 3rd person and it should still be black.

Note: This should not make the screen black if you're outside of a node due to the player collisionbox being significantly larger than the clipping distance of 1 cm (0.01 nodes).

### CODE REVIEW

File: src/client/clientmap.cpp [hunk 1]
```diff
@@ -677,12 +677,24 @@ void ClientMap::renderPostFx(CameraMode cam_mode)
...
+		}
+	}
+
 	video::SColor post_effect_color = features.post_effect_color;
 
 	// If the camera is in a solid node, make everything black.
-	// (first person mode only)
-	if (features.solidness == 2 && cam_mode == CAMERA_MODE_FIRST &&
-		!m_control.allow_noclip) {
+	if (any_nearby_solid && !m_control.allow_noclip) {
```
Comment (appgurueu) [inline src/client/clientmap.cpp:697]: Was there any reason for this being first person mode only? Surely the 3rd person mode isn't implemented *that hackily*, placing the camera inside solid nodes and hoping backface culling does all the work? Please test this extensively in 3rd person mode.

File: src/client/clientmap.cpp [hunk 2]
```diff
@@ -677,12 +677,24 @@ void ClientMap::renderPostFx(CameraMode cam_mode)
...
 
 	const ContentFeatures& features = m_nodedef->get(n);
+
+	// Account for the near clipping plane to counter see-through "glitches":
+	// Check all nodes within a radius of near_plane (Manhattan distance)
+	bool any_nearby_solid = false;
+	f32 near_plane = m_client->getCamera()->getCameraNode()->getNearValue();
+	for (s8 x = -1; x <= 1; x += 2) {
+		for (s8 y = -1; y <= 1; y += 2) {
+			for (s8 z = -1; z <= 1; z += 2) {
```
Comment (appgurueu) [inline src/client/clientmap.cpp:687]: This is a cartesian product ${-1, 1}^3$. Should I more compactly (and possibly more efficiently?) golf this to a single for loop using some bitpacking?

File: src/client/clientmap.cpp [hunk 3]
```diff
@@ -677,12 +677,24 @@ void ClientMap::renderPostFx(CameraMode cam_mode)
...
+				any_nearby_solid |= m_nodedef->get(getNode(floatToInt(
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/client/clientmap.cpp:688]: you could use `if` + `goto` to break out of the nested loop.

[the remark this answers is not in the document]
Comment (appgurueu) [reply src/client/clientmap.cpp:688]: Right, C++ doesn't allow naming loops.

File: src/client/clientmap.cpp (hunk 2 above)
Comment (sfan5) [reply src/client/clientmap.cpp:687]: We have `g_6dirs` for this, also you can skip this loop if `allow_noclip`.

File: src/client/clientmap.cpp (hunk 1 above)
Comment (sfan5) [reply src/client/clientmap.cpp:697]: If this correctly uses the position of the camera (not player) in 3rd person mode I don't see any issue with it.
Currently it's not too hard to get the 3P mode to look inside nodes like in the OP screenshot.

File: src/client/clientmap.cpp [hunk 4]
```diff
@@ -677,12 +677,24 @@ void ClientMap::renderPostFx(CameraMode cam_mode)
...
+						m_camera_position + near_plane * v3f(x, y, z), BS))).solidness == 2;
```
Comment (sfan5) [inline src/client/clientmap.cpp:689]: for readability please pull out the position into a separate assignment

File: src/client/clientmap.cpp (hunk 2 above)
Comment (appgurueu) [reply src/client/clientmap.cpp:687]: These are actually 8 dirs (the cuboid corners rather than the face centers).

Comment (sfan5) [reply src/client/clientmap.cpp:687]: oh huh
add it local to this file I guess, someone will hopefully move it to the right place if it's needed a second time

File: src/client/clientmap.cpp (hunk 1 above)
Comment (appgurueu) [reply src/client/clientmap.cpp:697]: I have to make this first-person only unfortunately because otherwise underground the screen is black way too often (e.g. when you're in a tunnel or the like). This pretty much defeats the purpose of the PR.

File: src/client/clientmap.cpp [hunk 5]
```diff
@@ -669,21 +669,37 @@ int ClientMap::getBackgroundBrightness(float max_d, u32 daylight_factor,
...
-	// (first person mode only)
-	if (features.solidness == 2 && cam_mode == CAMERA_MODE_FIRST &&
-		!m_control.allow_noclip) {
-		post_effect_color = video::SColor(255, 0, 0, 0);
+	video::SColor post_effect_color = m_nodedef->get(n).post_effect_color;
+
+	if (cam_mode == CAMERA_MODE_FIRST && !m_control.allow_noclip) {
+		// Account for the near clipping plane to counter see-through "glitches":
+		// Check all nodes within a radius of near_plane (Manhattan distance)
+		f32 near_plane = 1.5f * m_client->getCamera()->getCameraNode()->getNearValue();
```
Comment (appgurueu) [inline src/client/clientmap.cpp:694]: The factor of `1.5` I've had to introduce here seems rather arbitrary to me. Where did my reasoning go wrong, why is this needed? How can a node not be within a Manhattan distance of the near plane yet have it's face partially culled by the near plane?

**Comment by Wuzzy2** [comment]
Idea: Use black only by default but if `post_effect_color` is set, use that one instead.

File: src/client/clientmap.cpp [hunk 6]
```diff
@@ -669,21 +669,37 @@ int ClientMap::getBackgroundBrightness(float max_d, u32 daylight_factor,
 	return ret;
 }
 
+const v3f cuboid_corners[8] = {
```
Comment (sfan5) [inline src/client/clientmap.cpp:672]:
```suggestion
static const v3f cuboid_corners[8] = {
```

File: src/client/clientmap.cpp (hunk 5 above)
Comment (sfan5) [reply src/client/clientmap.cpp:694]: :shrug: but please mention the unclarity in a comment

**Comment by appgurueu** [comment]
Deprioritizing this in favor of my other three PRs. I don't really want to work on this further and it's too hacky in its current state.
